libnetwork/osl/neigh_linux.go

```diff
@@ -1,123 +1,126 @@
 package osl
 
 import (
-	"bytes"
 	"context"
 	"errors"
 	"fmt"
 	"net"
 	"os"
+	"strings"
 
 	"github.com/containerd/log"
 	"github.com/vishvananda/netlink"
 )
 
 // NeighborSearchError indicates that the neighbor is already present
 type NeighborSearchError struct {
-	ip      net.IP
-	mac     net.HardwareAddr
-	present bool
-}
-
-func (n NeighborSearchError) Error() string {
-	return fmt.Sprintf("Search neighbor failed for IP %v, mac %v, present in db:%t", n.ip, n.mac, n.present)
-}
-
-type neigh struct {
-	dstIP    net.IP
-	dstMac   net.HardwareAddr
+	ip       net.IP
+	mac      net.HardwareAddr
 	linkName string
-	linkDst  string
-	family   int
+	present  bool
 }
 
-func (n *Namespace) findNeighbor(dstIP net.IP, dstMac net.HardwareAddr) *neigh {
-	n.mu.Lock()
-	defer n.mu.Unlock()
-
-	for _, nh := range n.neighbors {
-		if nh.dstIP.Equal(dstIP) && bytes.Equal(nh.dstMac, dstMac) {
-			return nh
-		}
+func (n NeighborSearchError) Error() string {
+	var b strings.Builder
+	b.WriteString("neighbor entry ")
+	if n.present {
```

Comment:
This was the only part I had a comment on in the other PR. Non-blocking (from my perspective), but reposting here for others; https://github.com/moby/moby/pull/50081#discussion_r2108869390

> Not a blocker, but was mostly looking here for trade-offs between performance-optimisation (strings.Builder) and discoverability (being able to grep for the (partial) error message); wasn't sure how much this error is in a "hot path".

Reply:
I actually used the strings.Builder approach for readability rather than optimization. Note that there are two branches. To achieve something similar with `fmt.Sprintf` is not much more greppable. E.g.:
```go
s := fmt.Sprintf("neighbor entry %s for IP %v, mac %v", map[bool]string{true: "already exists", false: "not found"}, n.ip, n.mac)
if n.linkName != "" {
        s += fmt.Sprintf(", link %v", n.linkname)
}
return s
```